Mathlib/Data/Vector/MapLemmas.lean

```diff
@@ -0,0 +1,324 @@
+theorem mapAccumr₂_mapAccumr₂_left_right  (f₁ : γ → β → σ₁ → σ₁ × φ)
+                                          (f₂ : α → β → σ₂ → σ₂ × γ) :
+    (mapAccumr₂ f₁ (mapAccumr₂ f₂ xs ys s₂).snd ys s₁)
+    = let m := mapAccumr₂ (fun x y (s₁, s₂) =>
+                let r₂ := f₂ x y s₂
+                let r₁ := f₁ r₂.snd y s₁
+                ((r₁.fst, r₂.fst), r₁.snd)
+              )
+            xs ys (s₁, s₂);
+    (m.fst.fst, m.snd) := by
+  induction xs, ys using Vector.revInductionOn₂ generalizing s₁ s₂ <;> simp_all
+
+@[simp]
+theorem mapAccumr₂_mapAccumr₂_right_left  (f₁ : α → γ → σ₁ → σ₁ × φ)
+                                          (f₂ : α → β → σ₂ → σ₂ × γ) :
+    (mapAccumr₂ f₁ xs (mapAccumr₂ f₂ xs ys s₂).snd s₁)
+    = let m := mapAccumr₂ (fun x y (s₁, s₂) =>
+                let r₂ := f₂ x y s₂
+                let r₁ := f₁ x r₂.snd s₁
+                ((r₁.fst, r₂.fst), r₁.snd)
+              )
+            xs ys (s₁, s₂);
+    (m.fst.fst, m.snd) := by
+  induction xs, ys using Vector.revInductionOn₂ generalizing s₁ s₂ <;> simp_all
+
+@[simp]
+theorem mapAccumr₂_mapAccumr₂_right_right (f₁ : β → γ → σ₁ → σ₁ × φ)
+                                          (f₂ : α → β → σ₂ → σ₂ × γ) :
+    (mapAccumr₂ f₁ ys (mapAccumr₂ f₂ xs ys s₂).snd s₁)
+    = let m := mapAccumr₂ (fun x y (s₁, s₂) =>
+                let r₂ := f₂ x y s₂
+                let r₁ := f₁ y r₂.snd s₁
+                ((r₁.fst, r₂.fst), r₁.snd)
+              )
+            xs ys (s₁, s₂);
+    (m.fst.fst, m.snd) := by
+  induction xs, ys using Vector.revInductionOn₂ generalizing s₁ s₂ <;> simp_all
+
+
+end Binary
+
+
+end Fold
+
+
+/-!
+## Redundant state or input optimizations
+
+The following section are collection of rewrites to simplify, or even get rid, redundant
+accumulation state, or redundant inputs
+-/
+section RedundantState
+variable (xs : Vector α n) (ys : Vector β n)
+
+/--
+  If an accumulation function `f`, given an initial state `s`, produces `s` as its output state
+  for all possible input bits, then the state is redundant and can be optimized out
+-/
+@[simp]
+theorem mapAccumr_redundant_state (f : α → σ → σ × β) (s : σ) (h : ∀ a, (f a s).fst = s) :
+    mapAccumr f xs s = (s, (map (fun x => (f x s).snd) xs)) := by
+  clear ys
+  induction xs using revInductionOn <;> simp_all
+
+/--
+  If an accumulation function `f`, given an initial state `s`, produces `s` as its output state
+  for all possible input bits, then the state is redundant and can be optimized out
+-/
+@[simp]
+theorem mapAccumr₂_redundant_state (f : α → β → σ → σ × γ) (s : σ) (h : ∀ a b, (f a b s).fst = s) :
+    mapAccumr₂ f xs ys s = (s, (map₂ (fun x y => (f x y s).snd) xs ys)) := by
+  induction xs, ys using revInductionOn₂ <;> simp_all
+
+/-- If `f` takes a pair of states, but always returns the same value for both elements of the
+    pair, then we can simplify to just a single element of state
+  -/
+@[simp]
+theorem mapAccumr_redundant_pair (f : α → (σ × σ) → (σ × σ) × β)
```

Comment:
I feel like this theorem can be generalized. Are there more general conditions where you can replace the state type with a different state type, that proves this lemma as a special case?

Reply:
It probably can be, I'll have to think about it a bit more. I choose products in particular, because the earlier folding of nested `mapAccumr`s has a tendency to produce pairs with redundant state, so this theorem is needed to counterbalance that.

Reply:
The most general way to relate two applications of `mapAccumr`, with potentially different states, is bisimulation. I've added bisimulation lemmas, and proven this result in terms of bisimulation